✨ Update create-kind-cluster-with-SSL-passthrough.sh

clubanderson opened the pull request:
kflex does not work with a cluster named "kind-kubeflex" - it is expecting "kubeflex". Without this change, the instructions on the helm install page do not work properly. The problem comes with pulling the contexts using kflex

```
kubectl config delete-context its1 || true
kflex ctx its1
kubectl config delete-context wds1 || true
kflex ctx wds1
kflex ctx # switch back to the initial context
```

fails if "kind-kubeflex" is used

<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary

## Related issue(s)

Fixes #

MikeSpreitzer commented:
@clubanderson: what version of `kflex` are you using?

MikeSpreitzer commented:
@clubanderson: I am a little confused. The opening comment here speaks of cluster name, but the change in this PR concerns context name.

BTW, as far as I know, the release-0.23.0 quickstart is quite firm in its expectation that the hosting cluster is accessed through a context named "kind-kubeflex".

MikeSpreitzer commented:
Can you give a URL for the instructions that do not work?

MikeSpreitzer commented:
I tried looking at https://docs.kubestellar.io/release-0.23.0/direct/core-chart/ but I did not get a failure. My typescript is below.

```shell
mspreitz@mjs13 kubestellar % kind get clusters
No kind clusters found.
mspreitz@mjs13 kubestellar % 
mspreitz@mjs13 kubestellar % 
mspreitz@mjs13 kubestellar % bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v0.23.0/scripts/create-kind-cluster-with-SSL-passthrough.sh) --name kubeflex --port 9443
Creating "kubeflex" kind cluster with SSL passthrougn and 9443 port mapping...
No kind clusters found.
Creating cluster "kubeflex" ...
 ✓ Ensuring node image (kindest/node:v1.29.2) 🖼
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-kubeflex"
You can now use your cluster with:

kubectl cluster-info --context kind-kubeflex

Thanks for using kind! 😊
Installing an nginx ingress...
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
Pathcing nginx ingress to enable SSL passthrough...
deployment.apps/ingress-nginx-controller patched
Waiting for nginx ingress with SSL passthrough to be ready...
pod/ingress-nginx-controller-779f699d74-zcsrg condition met
Setting context to "kind-kubeflex"...
Switched to context "kind-kubeflex".
mspreitz@mjs13 kubestellar % 
mspreitz@mjs13 kubestellar % 
mspreitz@mjs13 kubestellar % kubectl config get-contexts | grep kubeflex
*         kind-kubeflex                                                                          kind-kubeflex                                    kind-kubeflex                                                                     
          wds2                                                                                   k3d-kubeflex                                     admin@k3d-kubeflex                                                                
mspreitz@mjs13 kubestellar % 
mspreitz@mjs13 kubestellar % 
mspreitz@mjs13 kubestellar % bash

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
bash-3.2$ 
bash-3.2$ 
bash-3.2$ 
bash-3.2$ export KUBESTELLAR_VERSION=0.23.0
bash-3.2$ 
bash-3.2$ helm upgrade --install ks-core oci://ghcr.io/kubestellar/kubestellar/core-chart --version $KUBESTELLAR_VERSION \
>   --set-json='ITSes=[{"name":"its1"}]' \
>   --set-json='WDSes=[{"name":"wds1"}]'
Release "ks-core" does not exist. Installing it now.
Pulled: ghcr.io/kubestellar/kubestellar/core-chart:0.23.0
Digest: sha256:550047534d8cca78020c7cb9907e760909a62021097183c6df277478bfa0d2bc
NAME: ks-core
LAST DEPLOYED: Wed Jun 26 17:16:13 2024
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
If you are using `kflex` from KubeFlex release 0.6.1 or earlier then
you should wait for the KubeFlex control planes (of type other than
"host") that you just made to become "ready". The following commands
will do that.

echo "Waiting for new non-host KubeFlex Control Planes to be Ready:"
for cpname in its1 wds1; do
  while [[ `kubectl get cp $cpname -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}'` != "True" ]]; do
    echo "Waiting for \"$cpname\"..."
    sleep 5
  done
  echo "\"$cpname\" is ready."
done

For your convenience you will probably want to add contexts to your
kubeconfig named after the non-host-type control planes (WDSes and
ITSes) that you just created (a host-type control plane is just an
alias for the KubeFlex hosting cluster). You can do that with the
following `kflex` commands; each creates a context and makes it the
current one. See
https://github.com/kubestellar/kubestellar/blob/main/docs/content/direct/core-chart.md#kubeconfig-files-and-contexts-for-control-planes
(replace "main" with "\{\{ .Values.KUBESTELLAR_VERSION \}\}" when
making the next release) for a way to do this without using `kflex`.


kubectl config delete-context its1 || true
kflex ctx its1
kubectl config delete-context wds1 || true
kflex ctx wds1

Finally you can use `kflex ctx` to switch back to the kubeconfig
context for your KubeFlex hosting cluster.
bash-3.2$ 
bash-3.2$ kflex version
Kubeflex version: v0.6.2.72af273 2024-05-29T01:38:44Z
Kubernetes version: v1.29.2
bash-3.2$ 
bash-3.2$ 
bash-3.2$ yq -e .preferences ~/.kube/config
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ 
bash-3.2$ echo "Waiting for new non-host KubeFlex Control Planes to be Ready:"
Waiting for new non-host KubeFlex Control Planes to be Ready:
bash-3.2$ for cpname in its1 wds1; do
>   while [[ `kubectl get cp $cpname -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}'` != "True" ]]; do
>     echo "Waiting for \"$cpname\"..."
>     sleep 5
>   done
>   echo "\"$cpname\" is ready."
> done
Waiting for "its1"...
Waiting for "its1"...
Waiting for "its1"...
Waiting for "its1"...
Waiting for "its1"...
Waiting for "its1"...
"its1" is ready.
"wds1" is ready.
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ 
bash-3.2$ kubectl config delete-context its1 || true
deleted context its1 from /Users/mspreitz/.kube/config
bash-3.2$ kflex ctx its1
kubeconfig context its1 not found (context its1 not found), trying to load from server...
✔ Switching to context its1...
bash-3.2$ kubectl config delete-context wds1 || true
deleted context wds1 from /Users/mspreitz/.kube/config
bash-3.2$ kflex ctx wds1
kubeconfig context wds1 not found (context wds1 not found), trying to load from server...
✔ Switching to context wds1...
bash-3.2$ 
bash-3.2$ kubectl config current-context
wds1
bash-3.2$ 
bash-3.2$ yq -e .preferences ~/.kube/config
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
bash-3.2$ kflex ctx
✔ Checking for saved hosting cluster context...
✔ Switching to hosting cluster context...
bash-3.2$ 
bash-3.2$ echo $?
0
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ 
```

clubanderson commented:
@MikeSpreitzer you did not use the instructions specified in the description above for removing and adding context which are directly from https://docs.kubestellar.io/release-0.23.0/direct/common-setup-core-chart/. You stopped short of that

ezrasilvera commented:
@clubanderson I remember that renaming the context had some side effects (e.g., on cleanup ) and maybe on some KF functionality (CC: @pdettori ).  But I see it passes all tests, I just can't recall the issue ...

MikeSpreitzer commented:
The above trial was not of the release branch. Below is a try using the release branch.

```shell
bash-3.2$ git status
On branch release-0.23.0
Your branch is up to date with 'upstream/release-0.23.0'.

nothing to commit, working tree clean
bash-3.2$ git rev-parse HEAD
85cfc0d973257930b7487bdb27699b6c18a4c26f
bash-3.2$ 
bash-3.2$ export KUBESTELLAR_VERSION=0.23.0
bash-3.2$ 
bash-3.2$ 
bash-3.2$ bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v0.23.0/scripts/create-kind-cluster-with-SSL-passthrough.sh) --name kubeflex --port 9443
Creating "kubeflex" kind cluster with SSL passthrougn and 9443 port mapping...
No kind clusters found.
Creating cluster "kubeflex" ...
 ✓ Ensuring node image (kindest/node:v1.29.2) 🖼
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-kubeflex"
You can now use your cluster with:

kubectl cluster-info --context kind-kubeflex

Have a nice day! 👋
Installing an nginx ingress...
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
Pathcing nginx ingress to enable SSL passthrough...
deployment.apps/ingress-nginx-controller patched
Waiting for nginx ingress with SSL passthrough to be ready...
pod/ingress-nginx-controller-779f699d74-tzm2x condition met
Setting context to "kind-kubeflex"...
Switched to context "kind-kubeflex".
bash-3.2$ 
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ yq -e .preferences ~/.kube/config
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
bash-3.2$ 
bash-3.2$ 
bash-3.2$ helm upgrade --install ks-core oci://ghcr.io/kubestellar/kubestellar/core-chart --version $KUBESTELLAR_VERSION \
>   --set-json='ITSes=[{"name":"its1"}]' \
>   --set-json='WDSes=[{"name":"wds1"}]'
Release "ks-core" does not exist. Installing it now.
Pulled: ghcr.io/kubestellar/kubestellar/core-chart:0.23.0
Digest: sha256:550047534d8cca78020c7cb9907e760909a62021097183c6df277478bfa0d2bc
NAME: ks-core
LAST DEPLOYED: Wed Jun 26 17:26:22 2024
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
If you are using `kflex` from KubeFlex release 0.6.1 or earlier then
you should wait for the KubeFlex control planes (of type other than
"host") that you just made to become "ready". The following commands
will do that.

echo "Waiting for new non-host KubeFlex Control Planes to be Ready:"
for cpname in its1 wds1; do
  while [[ `kubectl get cp $cpname -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}'` != "True" ]]; do
    echo "Waiting for \"$cpname\"..."
    sleep 5
  done
  echo "\"$cpname\" is ready."
done

For your convenience you will probably want to add contexts to your
kubeconfig named after the non-host-type control planes (WDSes and
ITSes) that you just created (a host-type control plane is just an
alias for the KubeFlex hosting cluster). You can do that with the
following `kflex` commands; each creates a context and makes it the
current one. See
https://github.com/kubestellar/kubestellar/blob/main/docs/content/direct/core-chart.md#kubeconfig-files-and-contexts-for-control-planes
(replace "main" with "\{\{ .Values.KUBESTELLAR_VERSION \}\}" when
making the next release) for a way to do this without using `kflex`.


kubectl config delete-context its1 || true
kflex ctx its1
kubectl config delete-context wds1 || true
kflex ctx wds1

Finally you can use `kflex ctx` to switch back to the kubeconfig
context for your KubeFlex hosting cluster.
bash-3.2$ 
bash-3.2$ 
bash-3.2$ yq -e .preferences ~/.kube/config
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ echo "Waiting for new non-host KubeFlex Control Planes to be Ready:"
Waiting for new non-host KubeFlex Control Planes to be Ready:
bash-3.2$ for cpname in its1 wds1; do
>   while [[ `kubectl get cp $cpname -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}'` != "True" ]]; do
>     echo "Waiting for \"$cpname\"..."
>     sleep 5
>   done
>   echo "\"$cpname\" is ready."
> done
Waiting for "its1"...
"its1" is ready.
"wds1" is ready.
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ yq -e .preferences ~/.kube/config
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ kubectl config delete-context its1 || true
deleted context its1 from /Users/mspreitz/.kube/config
bash-3.2$ kflex ctx its1
kubeconfig context its1 not found (context its1 not found), trying to load from server...
✔ Switching to context its1...
bash-3.2$ kubectl config delete-context wds1 || true
deleted context wds1 from /Users/mspreitz/.kube/config
bash-3.2$ kflex ctx wds1
kubeconfig context wds1 not found (context wds1 not found), trying to load from server...
✔ Switching to context wds1...
bash-3.2$ 
bash-3.2$ 
bash-3.2$ kubectl config current-context
wds1
bash-3.2$ yq -e .preferences ~/.kube/config
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
bash-3.2$ kflex ctx
✔ Checking for saved hosting cluster context...
✔ Switching to hosting cluster context...
bash-3.2$ echo $?
0
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ yq -e .preferences ~/.kube/config
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
bash-3.2$ kflex version
Kubeflex version: v0.6.2.72af273 2024-05-29T01:38:44Z
Kubernetes version: v1.29.2
```

clubanderson commented:
yep - the error is in your your typescript output

```
bash-3.2$ kubectl config delete-context its1 || true
deleted context its1 from /Users/mspreitz/.kube/config
bash-3.2$ kflex ctx its1
kubeconfig context its1 not found (context its1 not found), trying to load from server...
✔ Switching to context its1...
bash-3.2$ kubectl config delete-context wds1 || true
deleted context wds1 from /Users/mspreitz/.kube/config
bash-3.2$ kflex ctx wds1
kubeconfig context wds1 not found (context wds1 not found), trying to load from server...
✔ Switching to context wds1...
bash-3.2$ 
bash-3.2$ 
bash-3.2$ kubectl config current-context
```

MikeSpreitzer commented:
@clubanderson: Following is another trial, in which I ignore the NOTES printed from the Helm chart and read https://docs.kubestellar.io/release-0.23.0/direct/common-setup-core-chart/ instead. I also intersperse some commands that examine relevant state. I also `set -e` after a while to make sure that no failure is overlooked. I do not see a failure. When `kflex` prints "kubeconfig context wds1 not found (context wds1 not found), trying to load from server...", this is not a failure, it is a progress report.

```shell
bash-3.2$ git status
On branch release-0.23.0
Your branch is up to date with 'upstream/release-0.23.0'.

nothing to commit, working tree clean
bash-3.2$ 
bash-3.2$ git rev-parse HEAD
85cfc0d973257930b7487bdb27699b6c18a4c26f
bash-3.2$ 
bash-3.2$ yq -e .preferences ${KUBECONFIG:-$HOME/.kube/config}
null
Error: no matches found
bash-3.2$ echo ${KUBECONFIG:-$HOME/.kube/config}
/Users/mspreitz/.kube/config
bash-3.2$ 
bash-3.2$ kubectl config current-context
error: current-context is not set
bash-3.2$ 
bash-3.2$ 
bash-3.2$ export KUBESTELLAR_VERSION=0.23.0
bash-3.2$ export OCM_TRANSPORT_PLUGIN=0.1.11
bash-3.2$ 
bash-3.2$ 
bash-3.2$ bash <(curl -s https://raw.githubusercontent.com/kubestellar/kubestellar/v${KUBESTELLAR_VERSION}/scripts/create-kind-cluster-with-SSL-passthrough.sh) --name kubeflex --port 9443
Creating "kubeflex" kind cluster with SSL passthrougn and 9443 port mapping...
No kind clusters found.
Creating cluster "kubeflex" ...
 ✓ Ensuring node image (kindest/node:v1.29.2) 🖼
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-kubeflex"
You can now use your cluster with:

kubectl cluster-info --context kind-kubeflex

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
Installing an nginx ingress...
namespace/ingress-nginx created
serviceaccount/ingress-nginx created
serviceaccount/ingress-nginx-admission created
role.rbac.authorization.k8s.io/ingress-nginx created
role.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrole.rbac.authorization.k8s.io/ingress-nginx created
clusterrole.rbac.authorization.k8s.io/ingress-nginx-admission created
rolebinding.rbac.authorization.k8s.io/ingress-nginx created
rolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx created
clusterrolebinding.rbac.authorization.k8s.io/ingress-nginx-admission created
configmap/ingress-nginx-controller created
service/ingress-nginx-controller created
service/ingress-nginx-controller-admission created
deployment.apps/ingress-nginx-controller created
job.batch/ingress-nginx-admission-create created
job.batch/ingress-nginx-admission-patch created
ingressclass.networking.k8s.io/nginx created
validatingwebhookconfiguration.admissionregistration.k8s.io/ingress-nginx-admission created
Pathcing nginx ingress to enable SSL passthrough...
deployment.apps/ingress-nginx-controller patched
Waiting for nginx ingress with SSL passthrough to be ready...
pod/ingress-nginx-controller-779f699d74-6qtwf condition met
Setting context to "kind-kubeflex"...
Switched to context "kind-kubeflex".
bash-3.2$ echo $?
0
bash-3.2$ 
bash-3.2$ 
bash-3.2$ yq -e .preferences ${KUBECONFIG:-$HOME/.kube/config}
null
Error: no matches found
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ 
bash-3.2$ helm upgrade --install ks-core oci://ghcr.io/kubestellar/kubestellar/core-chart --version $KUBESTELLAR_VERSION \
>   --set-json='ITSes=[{"name":"its1"}]' \
>   --set-json='WDSes=[{"name":"wds1"}]'
Release "ks-core" does not exist. Installing it now.
Pulled: ghcr.io/kubestellar/kubestellar/core-chart:0.23.0
Digest: sha256:550047534d8cca78020c7cb9907e760909a62021097183c6df277478bfa0d2bc
NAME: ks-core
LAST DEPLOYED: Wed Jun 26 22:28:32 2024
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
If you are using `kflex` from KubeFlex release 0.6.1 or earlier then
you should wait for the KubeFlex control planes (of type other than
"host") that you just made to become "ready". The following commands
will do that.

echo "Waiting for new non-host KubeFlex Control Planes to be Ready:"
for cpname in its1 wds1; do
  while [[ `kubectl get cp $cpname -o 'jsonpath={..status.conditions[?(@.type=="Ready")].status}'` != "True" ]]; do
    echo "Waiting for \"$cpname\"..."
    sleep 5
  done
  echo "\"$cpname\" is ready."
done

For your convenience you will probably want to add contexts to your
kubeconfig named after the non-host-type control planes (WDSes and
ITSes) that you just created (a host-type control plane is just an
alias for the KubeFlex hosting cluster). You can do that with the
following `kflex` commands; each creates a context and makes it the
current one. See
https://github.com/kubestellar/kubestellar/blob/main/docs/content/direct/core-chart.md#kubeconfig-files-and-contexts-for-control-planes
(replace "main" with "\{\{ .Values.KUBESTELLAR_VERSION \}\}" when
making the next release) for a way to do this without using `kflex`.


kubectl config delete-context its1 || true
kflex ctx its1
kubectl config delete-context wds1 || true
kflex ctx wds1

Finally you can use `kflex ctx` to switch back to the kubeconfig
context for your KubeFlex hosting cluster.
bash-3.2$ echo $?
0
bash-3.2$ 
bash-3.2$ yq -e .preferences ${KUBECONFIG:-$HOME/.kube/config}
null
Error: no matches found
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ 
bash-3.2$ set -e
bash-3.2$ 
bash-3.2$ kubectl config delete-context its1 || true
deleted context its1 from /Users/mspreitz/.kube/config
bash-3.2$ 
bash-3.2$ kflex ctx its1
kubeconfig context its1 not found (context its1 not found), trying to load from server...
✔ Switching to context its1...
bash-3.2$ echo $?
0
bash-3.2$ 
bash-3.2$ kubectl config current-context
its1
bash-3.2$ 
bash-3.2$ 
bash-3.2$ yq -e .preferences ${KUBECONFIG:-$HOME/.kube/config}
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
bash-3.2$ kubectl config delete-context wds1 || true
deleted context wds1 from /Users/mspreitz/.kube/config
bash-3.2$ kflex ctx wds1
kubeconfig context wds1 not found (context wds1 not found), trying to load from server...
✔ Switching to context wds1...
bash-3.2$ echo $?
0
bash-3.2$ 
bash-3.2$ kubectl config current-context
wds1
bash-3.2$ 
bash-3.2$ yq -e .preferences ${KUBECONFIG:-$HOME/.kube/config}
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
bash-3.2$ kflex ctx # switch back to the initial context
◐ Switching to hosting cluster context...✔ Checking for saved hosting cluster context...
✔ Switching to hosting cluster context...
bash-3.2$ echo $?
0
bash-3.2$ 
bash-3.2$ kubectl --context its1 get deploy -A
NAMESPACE     NAME      READY   UP-TO-DATE   AVAILABLE   AGE
kube-system   coredns   1/1     1            1           91s
bash-3.2$ 
bash-3.2$ 
bash-3.2$ kubectl get deployments,statefulsets --all-namespaces
NAMESPACE            NAME                                             READY   UP-TO-DATE   AVAILABLE   AGE
ingress-nginx        deployment.apps/ingress-nginx-controller         1/1     1            1           6m34s
kube-system          deployment.apps/coredns                          2/2     2            2           6m36s
kubeflex-system      deployment.apps/kubeflex-controller-manager      1/1     1            1           4m56s
local-path-storage   deployment.apps/local-path-provisioner           1/1     1            1           6m35s
wds1-system          deployment.apps/kube-apiserver                   1/1     1            1           3m20s
wds1-system          deployment.apps/kube-controller-manager          1/1     1            1           3m20s
wds1-system          deployment.apps/kubestellar-controller-manager   0/1     1            0           96s
wds1-system          deployment.apps/transport-controller             0/1     1            0           97s

NAMESPACE         NAME                                   READY   AGE
its1-system       statefulset.apps/vcluster              1/1     3m15s
kubeflex-system   statefulset.apps/postgres-postgresql   1/1     4m22s
bash-3.2$ 
bash-3.2$ kubectl config current-context
kind-kubeflex
bash-3.2$ 
bash-3.2$ yq -e .preferences ${KUBECONFIG:-$HOME/.kube/config}
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
bash-3.2$ 
```

ezrasilvera commented:
@MikeSpreitzer I think you had some comments on the issue and proposed solution?

MikeSpreitzer commented:
@ezrasilvera: Indeed I commented on #2261 . I am still in the diagnostic phase, I have outstanding questions to @clubanderson .

MikeSpreitzer commented:
I think this PR was prompted by confusion generated by the fact that `kflex` uses some obscure state and in a fragile way. `kflex` uses a "preference" in the user's kubeconfig file that somehow got set to the wrong value, and `kflex` is reticent about changing that setting.

Here is an example of examining that obscure state:

```
mspreitz@mjs13 kubestellar % yq .preferences ${KUBECONFIG:-~/.kube/config}
extensions:
  - extension:
      data:
        kflex-initial-ctx-name: kind-kubeflex
      metadata:
        creationTimestamp: null
        name: kflex-config-extension-name
    name: kflex-config-extension-name
```

This shows that the relevant extension is present and is saying that the KubeFlex hosting cluster is to be accessed through the context named "kind-kubeflex".

I think that the OP somehow got this set to say "kubeflex" instead of "kind-kubeflex", and I suspect that`kflex` was reticent to change this setting even when the user started over from scratch and the setting should have been changed to say "kind-kubeflex" because that is actually the name of the context that `kflex` established for accessing the hosting cluster.

MikeSpreitzer commented:
As far as i can tell, release 0.6.2 of KubeFlex is NOT reticent to fix a wrong value for its extension in the user's kubeconfig file. Following is a demonstration.

```shell
mspreitz@mjs13 kubestellar % kflex version
Kubeflex version: v0.6.2.72af273 2024-05-29T01:38:44Z
Could not connect to a Kubernetes cluster: error building kubeconfig: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable

mspreitz@mjs13 kubestellar % grep kflex-initial-ctx-name ${KUBECONFIG:-$HOME/.kube/config}
        kflex-initial-ctx-name: wrong-answer

mspreitz@mjs13 kubestellar % kflex init -c
✔ Checking if OpenShift cluster...
...
✔ Installing kubeflex operator...
✔ Waiting for kubeflex operator to become ready...

mspreitz@mjs13 kubestellar % grep kflex-initial-ctx-name ${KUBECONFIG:-$HOME/.kube/config}
        kflex-initial-ctx-name: kind-kubeflex
```

MikeSpreitzer commented:
See Issue #2389, I think that is what remains from this PR.
